Jan24 patches

fuhlig1 opened the pull request:
- Fix the build systems of several packages to allow the usage of CMake 4.0.
- Fix compilation issues with Apple Clang 17
- Allow to use a non standard (older) SDK to allow compilation with Clang 17
       - Don't overwrite CMAKE_OSX_SYSROOT which is passed as CMake parameter
       - Pass CMAKE_OSX_SYSROOT to all package build systems
       - Fix ROOT build system such that rootcint takes into account the defined  CMAKE_OSX_SYSROOT

fuhlig1 committed:
[legacy] hepmc: Fix build system for CMake 4.0.1

fuhlig1 committed:
[legacy] zeromq: Fix build system for CMake 4.0.1

fuhlig1 committed:
[legacy] vc: Fix build system for CMake 4.0.1

fuhlig1 committed:
[legacy] pythia6: Fix build system for CMake 4.0.1

fuhlig1 committed:
[legacy] root: Fix compilation on macosx with Clang 17 and SDK 15

Clang 17 remove the header fp.h which is used in ROOT's copy of libAfterimage.
Add a patch which uses math.h instead.
The included ROOT version can't be compiled any longer with the SDK 15.
Specifing the older SDK 14 when running CMake (with -DCMAKE_OSX_SYSROOT)
only works partly since rootcling still uses the latest SDK version.
Add a patch such that rootcling also uses the specified SDK.
All projects depending on ROOT need also to set CMAKE_OSX_SYSROOT to the
same value used when building ROOT. Otherwise the dictionary generation
may fail.

Fix patch for ROOT using CMAKE_OSX_SYSROOT

minimize the patch

fuhlig1 committed:
[legacy] build system:  Allow setting the proper SDK

Before a SDK defined via the CMake parameter CMAKE_OSX_SYSROOT was
overwritten with the value extracted from brew.

fuhlig1 committed:
[legacy] geant4: Fix compilation crash with Clang 17

Fix a typo in the code. This code could actually never be compiled and
it is unclear why this compilation error only popped up now with
Clang 17.

fuhlig1 committed:
[legacy] geant4: Fix config file

The fix prevents a crash about wrong CMake versions when compiling
geant4_vmc.

fuhlig1 committed:
[legacy] vmc: Fix build system for CMake 4.0.1

fuhlig1 committed:
[legacy] vgm: Fix build system for CMake 4.0.1

fuhlig1 committed:
[legacy] geant4_vmc: Fix build system for CMake 4.0.1

fuhlig1 committed:
[legacy] geant3: Fix build system for CMake 4.0.1

Also fix an issue with the usage of the pythia6 library. The libarary
wasn't found any longer such that it is explicitly added as
dependency.

fuhlig1 committed:
[legacy] setup scripts: Add new packages for macosx

The default bash version on macosx is to old to support all features
needed in the test suite of FairRoot.
The system version of make doesn't properly support the jobserver
such that the build process ins't properly parallelized.

fuhlig1 committed:
[legacy] documentation: Add infos macosx

Describe some known problems on macosx and the corresponding
solutions.

dennisklein commented on legacy/root/fix_macos_sdk_for_rootcling.patch:
```diff
@@ -0,0 +1,119 @@
+-      ${CMAKE_SOURCE_DIR}/build/unix/makepchinput.py
+-      ${CMAKE_SOURCE_DIR}/etc/dictpch/makepch.py
+-  )
++  if(APPLE AND CMAKE_OSX_SYSROOT)
++    add_custom_command(OUTPUT etc/allDict.cxx.pch
++      BYPRODUCTS
++        etc/dictpch/allCppflags.txt
++        etc/dictpch/allHeaders.h
++        etc/dictpch/allLinkDefs.h
++      COMMAND
++        ${PYTHON_EXECUTABLE} ${CMAKE_SOURCE_DIR}/build/unix/makepchinput.py
++        ${CMAKE_SOURCE_DIR} . ${pyroot_legacy} ${__cling_pch}
++      COMMAND
++        ${CMAKE_COMMAND} -E env "SDKROOT=${CMAKE_OSX_SYSROOT}" env ROOTIGNOREPREFIX=1 ${PYTHON_EXECUTABLE}
++        ${CMAKE_SOURCE_DIR}/etc/dictpch/makepch.py etc/allDict.cxx.pch
++        ${__allIncludes} -I${CMAKE_BINARY_DIR}/include -I${CMAKE_SOURCE_DIR}/core
++      DEPENDS
++        rootcling ${__pch_dependencies} ${__pch_dictionaries}
++        ${CMAKE_SOURCE_DIR}/build/unix/makepchinput.py
++        ${CMAKE_SOURCE_DIR}/etc/dictpch/makepch.py
++    )
++  else()
++    add_custom_command(OUTPUT etc/allDict.cxx.pch
++      BYPRODUCTS
++        etc/dictpch/allCppflags.txt
++        etc/dictpch/allHeaders.h
++        etc/dictpch/allLinkDefs.h
++      COMMAND
++        ${PYTHON_EXECUTABLE} ${CMAKE_SOURCE_DIR}/build/unix/makepchinput.py
++        ${CMAKE_SOURCE_DIR} . ${pyroot_legacy} ${__cling_pch}
++      COMMAND
++        ${CMAKE_COMMAND} -E env ROOTIGNOREPREFIX=1 ${PYTHON_EXECUTABLE}
++        ${CMAKE_SOURCE_DIR}/etc/dictpch/makepch.py etc/allDict.cxx.pch
++        ${__allIncludes} -I${CMAKE_BINARY_DIR}/include -I${CMAKE_SOURCE_DIR}/core
++      DEPENDS
++        rootcling ${__pch_dependencies} ${__pch_dictionaries}
++        ${CMAKE_SOURCE_DIR}/build/unix/makepchinput.py
++        ${CMAKE_SOURCE_DIR}/etc/dictpch/makepch.py
++    )
++  endif()
```
For my taste this is a lot of code duplication, only one line adds a new argument?!

fuhlig1 replied:
Will try it. I think the problem was the double quotes which were not properly handled but maybe your version works.

fuhlig1 commented on legacy/README.md:
````diff
@@ -51,6 +51,53 @@ cmake -S <path-to-source> -B <path-to-build> -C <path-to-source>/FairSoftConfig.
+To allow CMake to find the proper version it is nedded to modify the PATH environment
+variable when running cmake
+
+```
+PATH=/usr/local/opt/make/libexec/gnubin:/usr/local/opt/gpatch/libexec/gnubin:$PATH cmake <all your parameters>
+```
````
Sounds good. I will try if this works for gmake and gpatch.

fuhlig1 committed:
[legacy] On macosx use Homebrews patch and make

fuhlig1 commented on cmake/legacy.cmake:
```diff
@@ -45,9 +45,31 @@ endif()
+  else()
+    message(STATUS "Found Make: ${_make}")
+  endif()
+  set(CMAKE_MAKE_PROGRAM ${_make} CACHE FILEPATH "Make program" FORCE)
```
In principle it is already set. That's why I check for a path ending with "/make". This should catch the case if /usr/bin/make is already set. All other tools shouldn't be affected but I still have to test when I explicitly use Ninja.

fuhlig1 replied:
Ok. After quite some trial and error I turned out that setting it in FairSoftConfig.cmake interferes with a generator passed by -G on the command line. The current code in cmake/legacy.cmake works correctly. Keep it the way it is.

fuhlig1 commented on legacy/geant3/fix_cmake.patch:
```diff
@@ -0,0 +1,37 @@
+ 
+ #-- VMC (required) ------------------------------------------------------------
+ find_package(VMC CONFIG REQUIRED)
++find_package(Pythia6 REQUIRED)
++set(VMC_DEPS VMCLibrary ${Pythia6_LIBRARY})
```
Add more info to the commit message.

fuhlig1 committed:
[legacy] boost: Build Boost with correct SDK

Pass the needed information to the boost build system if a non
standard SDK is used.

fuhlig1 committed:
[legacy] pythia8: Build Pythia8 with correct SDK

Pass the needed information to the pythia8 build system if a non
standard SDK is used.

fuhlig1 committed:
[legacy] root: Fix build with gcc15

Some legacy C code can't be compiled with C standard version 23 which
is the new default for gcc15 so use C standard version 11 for the
compilation.
Add some missing include statements for <stdint>. Probably due to
header cleanup in the include files of gcc 15.

fuhlig1 committed:
[legacy] geant3: Fix compilation for gcc15

The legacy C code can't be compiled with C standard version 23 which
is the new default for gcc15 so use C standard version 11 for the
compilation.
For unknown reasons the pythia library isn't found any longer which
is fixed by adding possible library names.

fuhlig1 committed:
[legacy] tests: Fix test for gcc15

Use C11 as standard C version when compiling FairRoot. The C code
of the MbsApi can't be compiled with C23 which is the new default
version with gcc15.

fuhlig1 committed:
[legacy] root: Fix compilation for some systems

The version of ftgl contained in the ROOT source code results in a
compilation error with the gcc14 compiler suite used in Debian 13 and
the gcc15 compiler suite in OpenSuse 16.0.
For all other tested system the code is okay.
The problem was also found by the Debian team

https://www.mail-archive.com/debian-qa-packages@lists.debian.org/msg81745.html

and a fix is available from upstream

https://github.com/frankheckenbach/ftgl/pull/20

Add a cmake check for OS and apply the fix only in case a Debian 13 or
OpenSuse 16.0 is found.

fuhlig1 committed:
[legacy] test: Allow test with non standard SDK on macosx

Properly pass CMAKE_OSX_SYSROOT to the build and test processes.
Don't test building onnxruntime if CMake version >= 4.0.0

fuhlig1 committed:
[legacy] test: Add test of FairRoot dev branch
